cip/1.accepted/CIP2017-02-07-Map-Projection.adoc

```diff
@@ -0,0 +1,121 @@
+• *Literal entry* - includes a entry into the projected map using the regular map entry syntax, i.e. with the value being explicitly different from the key.
+  In this case the value can be any arbitrary expression, just like in a regular map literal.
+• *All properties selector* - includes all properties from the original map or entity into the projected map. +
+  The all properties selector also allows marking properties for exclusion that would otherwise have been included by the selector. +
+  If any properties explicitly stated in the map projection through another selector type conflicts with a property in the original map, the explicitly stated entry takes precedence and the value from that expression overrides the value from the original map.
+
+Note that while it is typically superfluous to use both an _all properties selector_ and explicit _property selectors_, this is syntactically allowed.
+Sometimes this might be the most convenient way to express the desired projection, since the explicit _property selector_ allows further projection of the nested property, and this value takes precedence over the _all properties selector_.
+
+=== Syntax
+
+The EBNF syntax of map projection takes the following place in the Cypher syntax:
+
+[source, ebnf]
+----
+Atom = … | MapProjection | … ;
+
+MapProjection = Variable, ProjectedMap ;
+
+ProjectedMap = '{', (AllPropertiesSelector | ProjectionEntry), {',', ProjectionEntry}, '}' ;
+
+ProjectionEntry = LiteralEntry
+                | PropertySelector
+                | VariableSelector
+                ;
+LiteralEntry = PropertyKeyName, ':', Expression ;
+PropertySelector = '.', PropertyKeyName, [ProjectedMap] ;
+VariableSelector = Variable, [ProjectedMap] ;
+AllPropertiesSelector = '*', {'-', PropertyKeyName} ;
+
+ReturnItem = Expression
+           | (Expression, 'AS', Variable)
+           | (ProjectedMap, 'AS', Variable)
+           ;
+----
+
+=== Examples
+
+[source, cypher]
+.Fetch name and address for a person
+----
+MATCH (person:Person{userId=$user})-[:ADDRESS]->(address)
+RETURN person {.firstName, .lastName, id: $user,
+               address {.streetAddress, .city, .postalCode} }
+----
+
+.Example result
+----
++----------------------------------------------------------------------------------------+
+| person                                                                                 |
++----------------------------------------------------------------------------------------+
+| {firstName: "Sherlock", lastName: "Holmes", id: "0099CC",                              |
+|  address: {streetAddress: "221B Baker Street", city: "London", postalCode: "NW1 6XE"}} |
++----------------------------------------------------------------------------------------+
+----
+
+[source, cypher]
+.Bind variables into a map
+----
+MATCH (group:Group{name:$groupName})<-[:MEMBER_OF]-(member)-[:KNOWS]-(friend)-[:ADDRESS]->(address)
```

Comment:
Cypher style: space between label and property predicate.
Cypher style: space after `:` in property predicate.